File: backend/internal/models/lab_session.go
package models

import (
	"fmt"
	"math"
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
)

// Lab session status constants.
const (
	StatusPending      = "pending"
	StatusInitializing = "initializing"
	StatusRunning      = "running"
	StatusStopping     = "stopping"
	StatusStopped      = "stopped"
	StatusTerminated   = "terminated"
	StatusError        = "error"
)

// Lab type constants.
const (
	LabTypeWebExploitation     = "web_exploitation"
	LabTypeNetworkPentesting   = "network_pentesting"
	LabTypePrivilegeEscalation = "privilege_escalation"
	LabTypeMalwareAnalysis     = "malware_analysis"
	LabTypeForensics           = "forensics"
	LabTypeCTFChallenge        = "ctf_challenge"
	LabTypeCustom              = "custom"
)

// StatusHistoryEntry represents a single status change event.
type StatusHistoryEntry struct {
	Status    string    `bson:"status" json:"status"`
	Timestamp time.Time `bson:"timestamp" json:"timestamp"`
	Message   string    `bson:"message,omitempty" json:"message,omitempty"`
}

// AccessCredentials holds lab session access credentials.
type AccessCredentials struct {
	Username string `bson:"username,omitempty" json:"username,omitempty"`
	Password string `bson:"password,omitempty" json:"-"`
}

// Checkpoint represents a completed objective in a lab.
type Checkpoint struct {
	CheckpointID string    `bson:"checkpointId" json:"checkpointId"`
	CompletedAt  time.Time `bson:"completedAt" json:"completedAt"`
	Points       int       `bson:"points" json:"points"`
}

// Flag represents a captured flag in a lab.
type Flag struct {
	FlagID  string    `bson:"flagId" json:"flagId"`
	Flag    string    `bson:"flag" json:"flag"`
	FoundAt time.Time `bson:"foundAt" json:"foundAt"`
	Points  int       `bson:"points" json:"points"`
}

// ActivityLogEntry represents an activity event in a session.
type ActivityLogEntry struct {
	Action    string      `bson:"action" json:"action"`
	Timestamp time.Time   `bson:"timestamp" json:"timestamp"`
	Details   interface{} `bson:"details,omitempty" json:"details,omitempty"`
}

// LabSession represents a user's lab session in MongoDB.
type LabSession struct {
	ID               bson.ObjectID      `bson:"_id,omitempty" json:"id"`
	User             bson.ObjectID      `bson:"user" json:"user"`
	LabType          string             `bson:"labType" json:"labType"`
	LabName          string             `bson:"labName" json:"labName"`
	Description      string             `bson:"description,omitempty" json:"description,omitempty"`
	Difficulty       string             `bson:"difficulty" json:"difficulty"`
	Status           string             `bson:"status" json:"status"`
	StatusHistory    []StatusHistoryEntry `bson:"statusHistory,omitempty" json:"statusHistory,omitempty"`

	// Docker container information (Phase 1.2)
	ContainerID  string `bson:"containerId,omitempty" json:"containerId,omitempty"`
	DockerImage  string `bson:"dockerImage,omitempty" json:"dockerImage,omitempty"`

	// AWS EC2 Information (Phase 6 — kept for future AWS migration)
	AWSInstanceID      string `bson:"awsInstanceId,omitempty" json:"awsInstanceId,omitempty"`
	AWSInstanceType    string `bson:"awsInstanceType" json:"awsInstanceType"`
	AWSAmiID           string `bson:"awsAmiId,omitempty" json:"awsAmiId,omitempty"`
	AWSSecurityGroupID string `bson:"awsSecurityGroupId,omitempty" json:"awsSecurityGroupId,omitempty"`
	AWSSubnetID        string `bson:"awsSubnetId,omitempty" json:"awsSubnetId,omitempty"`
	AWSRegion          string `bson:"awsRegion" json:"awsRegion"`

	// Connection Details
	PublicIP          string            `bson:"publicIp,omitempty" json:"publicIp,omitempty"`
	PrivateIP         string            `bson:"privateIp,omitempty" json:"privateIp,omitempty"`
	SSHKeyName        string            `bson:"sshKeyName,omitempty" json:"sshKeyName,omitempty"`
	AccessCredentials AccessCredentials `bson:"accessCredentials,omitempty" json:"accessCredentials,omitempty"`

	// Time Tracking
	StartTime           *time.Time `bson:"startTime,omitempty" json:"startTime,omitempty"`
	EndTime             *time.Time `bson:"endTime,omitempty" json:"endTime,omitempty"`
	TotalBillableMinutes int       `bson:"totalBillableMinutes" json:"totalBillableMinutes"`
	MaxDuration         int        `bson:"maxDuration" json:"maxDuration"` // in minutes
	AutoTerminateAt     *time.Time `bson:"autoTerminateAt,omitempty" json:"autoTerminateAt,omitempty"`
	WarningNotificationSent bool   `bson:"warningNotificationSent" json:"warningNotificationSent"`

	// Cost Tracking
	HourlyRate    float64  `bson:"hourlyRate" json:"hourlyRate"`
	EstimatedCost float64  `bson:"estimatedCost" json:"estimatedCost"`
	FinalCost     *float64 `bson:"finalCost,omitempty" json:"finalCost,omitempty"`

	// Progress
	CheckpointsCompleted []Checkpoint     `bson:"checkpointsCompleted,omitempty" json:"checkpointsCompleted,omitempty"`
	FlagsFound           []Flag           `bson:"flagsFound,omitempty" json:"flagsFound,omitempty"`
	Score                int              `bson:"score" json:"score"`

	// Logs and Notes
	ActivityLog []ActivityLogEntry `bson:"activityLog,omitempty" json:"activityLog,omitempty"`
	UserNotes   string             `bson:"userNotes,omitempty" json:"userNotes,omitempty"`

	// Error Information
	ErrorMessage string      `bson:"errorMessage,omitempty" json:"errorMessage,omitempty"`
	ErrorDetails interface{} `bson:"errorDetails,omitempty" json:"errorDetails,omitempty"`

	// Metadata
	Metadata interface{} `bson:"metadata,omitempty" json:"metadata,omitempty"`

	CreatedAt time.Time `bson:"createdAt" json:"createdAt"`
	UpdatedAt time.Time `bson:"updatedAt" json:"updatedAt"`
}

// DurationMinutes calculates the session duration in minutes.
func (s *LabSession) DurationMinutes() int {
	if s.StartTime == nil {
		return 0
	}
	endTime := time.Now()
	if s.EndTime != nil {
		endTime = *s.EndTime
	}
	return int(math.Ceil(endTime.Sub(*s.StartTime).Minutes()))
}

// DurationFormatted returns a human-readable duration string.
func (s *LabSession) DurationFormatted() string {
	minutes := s.DurationMinutes()
	hours := minutes / 60
	mins := minutes % 60
	if hours > 0 {
		return fmt.Sprintf("%dh %dm", hours, mins)
	}
	return fmt.Sprintf("%dm", mins)
}

// CurrentCost calculates the real-time cost.
func (s *LabSession) CurrentCost() float64 {
	minutes := float64(s.DurationMinutes())
	hours := minutes / 60.0
	cost := hours * s.HourlyRate
	return math.Round(cost*100) / 100
}

// IsActive returns true if the session is in an active state.
func (s *LabSession) IsActive() bool {
	return s.Status == StatusPending || s.Status == StatusInitializing || s.Status == StatusRunning
}

// ValidStatusTransitions maps current status to allowed next statuses.
var ValidStatusTransitions = map[string][]string{
	StatusPending:      {StatusInitializing, StatusError},
	StatusInitializing: {StatusRunning, StatusError},
	StatusRunning:      {StatusStopping, StatusTerminated, StatusError},
	StatusStopping:     {StatusStopped, StatusError},
	StatusStopped:      {StatusRunning, StatusTerminated},
}

// IsValidTransition checks if a status transition is allowed.
func IsValidTransition(from, to string) bool {
	allowed, exists := ValidStatusTransitions[from]
	if !exists {
		return false
	}
	for _, s := range allowed {
		if s == to {
			return true
		}
	}
	return false
}

File: server/src/models/LabSession.js
import mongoose from "mongoose";
import config from "../config/index.js";

/**
 * Lab Session Status Flow:
 *
 * PENDING -> INITIALIZING -> RUNNING -> STOPPING -> STOPPED -> TERMINATED
 *                                    -> ERROR (can happen at any stage)
 *
 * PENDING: User requested lab, waiting for system
 * INITIALIZING: AWS EC2 instance is being provisioned
 * RUNNING: Instance is active, user can access
 * STOPPING: User/system initiated stop, instance stopping
 * STOPPED: Instance stopped but not terminated (can be restarted)
 * TERMINATED: Instance terminated, session complete
 * ERROR: Something went wrong during any stage
 */
export const LAB_STATUS = {
  PENDING: "pending",
  INITIALIZING: "initializing",
  RUNNING: "running",
  STOPPING: "stopping",
  STOPPED: "stopped",
  TERMINATED: "terminated",
  ERROR: "error",
};

/**
 * Lab Types for different training scenarios
 */
export const LAB_TYPES = {
  WEB_EXPLOITATION: "web_exploitation",
  NETWORK_PENTESTING: "network_pentesting",
  PRIVILEGE_ESCALATION: "privilege_escalation",
  MALWARE_ANALYSIS: "malware_analysis",
  FORENSICS: "forensics",
  CTF_CHALLENGE: "ctf_challenge",
  CUSTOM: "custom",
};

const labSessionSchema = new mongoose.Schema(
  {
    // User Reference
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: [true, "User reference is required"],
      index: true,
    },

    // Lab Configuration
    labType: {
      type: String,
      enum: Object.values(LAB_TYPES),
      default: LAB_TYPES.WEB_EXPLOITATION,
    },
    labName: {
      type: String,
      required: [true, "Lab name is required"],
      trim: true,
      maxlength: [100, "Lab name cannot exceed 100 characters"],
    },
    description: {
      type: String,
      trim: true,
      maxlength: [500, "Description cannot exceed 500 characters"],
    },
    difficulty: {
      type: String,
      enum: ["beginner", "intermediate", "advanced", "expert"],
      default: "beginner",
    },

    // Session Status
    status: {
      type: String,
      enum: Object.values(LAB_STATUS),
      default: LAB_STATUS.PENDING,
      index: true,
    },
    statusHistory: [
      {
        status: {
          type: String,
          enum: Object.values(LAB_STATUS),
        },
        timestamp: {
          type: Date,
          default: Date.now,
        },
        message: String,
      },
    ],

    // AWS EC2 Information (Phase 2+)
    awsInstanceId: {
      type: String,
      default: null,
      index: true,
    },
    awsInstanceType: {
      type: String,
      default: "t2.micro", // Free tier eligible
    },
    awsAmiId: {
      type: String,
      default: null,
    },
    awsSecurityGroupId: {
      type: String,
      default: null,
    },
    awsSubnetId: {
      type: String,
      default: null,
    },
    awsRegion: {
      type: String,
      default: config.aws.region || "us-east-1",
    },

    // Connection Details (Phase 2+)
    publicIp: {
      type: String,
      default: null,
    },
    privateIp: {
      type: String,
      default: null,
    },
    sshKeyName: {
      type: String,
      default: null,
    },
    accessCredentials: {
      username: String,
      // Password stored encrypted in production
      password: {
        type: String,
        select: false,
      },
    },

    // Time Tracking (Critical for billing)
    startTime: {
      type: Date,
      default: null,
    },
    endTime: {
      type: Date,
      default: null,
    },
    // Billable time in minutes
    totalBillableMinutes: {
      type: Number,
      default: 0,
    },
    // Auto-terminate settings
    maxDuration: {
      type: Number,
      default: 240, // 4 hours in minutes
    },
    autoTerminateAt: {
      type: Date,
      default: null,
    },
    warningNotificationSent: {
      type: Boolean,
      default: false,
    },

    // Cost Tracking (@ $0.50/hour)
    hourlyRate: {
      type: Number,
      default: config.lab.hourlyRate || 0.5,
    },
    estimatedCost: {
      type: Number,
      default: 0,
    },
    finalCost: {
      type: Number,
      default: null,
    },

    // User Progress/Activity
    checkpointsCompleted: [
      {
        checkpointId: String,
        completedAt: Date,
        points: Number,
      },
    ],
    flagsFound: [
      {
        flagId: String,
        flag: String,
        foundAt: Date,
        points: Number,
      },
    ],
    score: {
      type: Number,
      default: 0,
    },

    // Logs and Notes
    activityLog: [
      {
        action: String,
        timestamp: {
          type: Date,
          default: Date.now,
        },
        details: mongoose.Schema.Types.Mixed,
      },
    ],
    userNotes: {
      type: String,
      maxlength: [5000, "Notes cannot exceed 5000 characters"],
    },

    // Error Information
    errorMessage: {
      type: String,
      default: null,
    },
    errorDetails: {
      type: mongoose.Schema.Types.Mixed,
      default: null,
    },

    // Metadata
    metadata: {
      type: mongoose.Schema.Types.Mixed,
      default: {},
    },
  },
  {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  }
);

// Compound indexes for common queries
labSessionSchema.index({ user: 1, status: 1 });
labSessionSchema.index({ user: 1, createdAt: -1 });
labSessionSchema.index({ status: 1, autoTerminateAt: 1 });
// Note: awsInstanceId index created automatically by index: true in field definition

// Virtual for session duration in minutes
labSessionSchema.virtual("durationMinutes").get(function () {
  if (!this.startTime) return 0;

  const endTime = this.endTime || new Date();
  const diffMs = endTime - this.startTime;
  return Math.ceil(diffMs / (1000 * 60));
});

// Virtual for formatted duration
labSessionSchema.virtual("durationFormatted").get(function () {
  const minutes = this.durationMinutes;
  const hours = Math.floor(minutes / 60);
  const mins = minutes % 60;

  if (hours > 0) {
    return `${hours}h ${mins}m`;
  }
  return `${mins}m`;
});

// Virtual for current cost (real-time calculation)
labSessionSchema.virtual("currentCost").get(function () {
  const minutes = this.durationMinutes;
  const hours = minutes / 60;
  return parseFloat((hours * this.hourlyRate).toFixed(2));
});

// Virtual to check if session is active
labSessionSchema.virtual("isActive").get(function () {
  return ["pending", "initializing", "running"].includes(this.status);
});

// Pre-save middleware to update status history
labSessionSchema.pre("save", function (next) {
  // Track status changes
  if (this.isModified("status")) {
    this.statusHistory.push({
      status: this.status,
      timestamp: new Date(),
      message: `Status changed to ${this.status}`,
    });
  }

  // Set startTime when status becomes 'running'
  if (
    this.isModified("status") &&
    this.status === LAB_STATUS.RUNNING &&
    !this.startTime
  ) {
    this.startTime = new Date();

    // Calculate auto-terminate time
    if (this.maxDuration) {
      this.autoTerminateAt = new Date(
        Date.now() + this.maxDuration * 60 * 1000
      );
    }
  }

  // Set endTime when session ends
  if (
    this.isModified("status") &&
    [LAB_STATUS.STOPPED, LAB_STATUS.TERMINATED].includes(this.status)
  ) {
    if (!this.endTime) {
      this.endTime = new Date();
    }

    // Calculate final cost
    if (this.startTime) {
      const minutes = Math.ceil((this.endTime - this.startTime) / (1000 * 60));
      this.totalBillableMinutes = minutes;
      this.finalCost = parseFloat(
        ((minutes / 60) * this.hourlyRate).toFixed(2)
      );
    }
  }

  next();
});

// Instance method to add activity log
labSessionSchema.methods.logActivity = function (action, details = {}) {
  this.activityLog.push({
    action,
    timestamp: new Date(),
    details,
  });
  return this.save();
};

// Instance method to update status
labSessionSchema.methods.updateStatus = function (newStatus, message = "") {
  this.status = newStatus;
  if (message) {
    this.statusHistory[this.statusHistory.length - 1].message = message;
  }
  return this.save();
};

// Static method to find active sessions for a user
labSessionSchema.statics.findActiveByUser = function (userId) {
  return this.find({
    user: userId,
    status: {
      $in: [LAB_STATUS.PENDING, LAB_STATUS.INITIALIZING, LAB_STATUS.RUNNING],
    },
  }).sort({ createdAt: -1 });
};

// Static method to find sessions that need auto-termination
labSessionSchema.statics.findForAutoTermination = function () {
  return this.find({
    status: LAB_STATUS.RUNNING,
    autoTerminateAt: { $lte: new Date() },
  });
};

// Static method to get user's total spent
labSessionSchema.statics.getUserTotalSpent = async function (userId) {
  const result = await this.aggregate([
    { $match: { user: userId, finalCost: { $ne: null } } },
    { $group: { _id: null, total: { $sum: "$finalCost" } } },
  ]);
  return result.length > 0 ? result[0].total : 0;
};

const LabSession = mongoose.model("LabSession", labSessionSchema);

export default LabSession;
